### ISSUE TITLE
Fortran compile fails with module dependencies

### ISSUE THREAD

**Issue by james-thunes** [SConsProject/scons#4177]
**Describe the bug**
scons does not appear to be correctly find dependencies on *.mod files in Fortran files. 

My project includes a Fortran library that includes a number of source files and a module including some shared variables. When compiling on windows with ifort, I get the following error when doing a clean build:
`error #7002: Error in opening the compiled module file.  Check INCLUDE paths. `
Subsequent builds complete successfully.

Investigation shows that the above compiler error is seen because the *.mod file required by the source files is not built before the source file compilation. Reordering the list of source files improves the situation, but it still fails periodically when building in parallel. Subsequent attempts to compile work as the *.mod file is created during the first scons attempt and are thus already available.

Per suggestion from discord, I ran scons with `--tree=prune` to check if the consumers of the module correctly listed the *.mod file as a build dependency. The output below (anonymized) shows the output for one of the source files:
```
  |   +-<path>\<objName>.staticrt.obj
  |   | +-<relativePath>\<srcFile>.f
  |   | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
```
Note that the source file does not list the *.mod file as a dependency.

The project is cross-platform so I also checked the result of `--tree=prune` on linux (compiled with gfortran). Scons also did not recognize the dependency on the mod file (output was the same as above with the exception of the compiler path). However, the code compiles without issue on linux.

Since the above shows that the source files are not identifying the *.mod file as a build dependency, my assumption is that there is some issue in the way that scons is determining dependencies for Fortran files. I'm not sure why there's not an issue with gfortran on linux.

**Required information**
* Link to SCons Users thread discussing your issue.
   discussed on the scons discord (fortran-support channel)
* Version of SCons
   3.1.2
* Version of Python
   3.7.13
* Which python distribution if applicable (python.org, cygwin, anaconda, macports, brew,etc)
   cpython
* How you installed SCons
   pip install
* What Platform are you on? (Linux/Windows and which version)
   Windows 10 Enterprise version1909
* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues.
* How you invoke scons (The command line you're using "scons --flags some_arguments")
   scons -j8

**Comment by james-thunes** [comment]
I haven't been able to get a small test case set up yet. I will try to create one and post to this issue

**Comment by dnwillia-work** [comment]
I'll just add that when I was testing this locally it seems to only be reproducible when running a parallel build with `-jN`.  A serial build would do things in the right order and I never saw that fail, but for parallel it's failing when a dependent source file is compiled concurrently before the module is complete.

**Comment by bdbaddog** [comment]
> I'll just add that when I was testing this locally it seems to only be reproducible when running a parallel build with `-jN`. A serial build would do things in the right order and I never saw that fail, but for parallel it's failing when a dependent source file is compiled concurrently before the module is complete.

That's just luck... If parallel builds fail due to missing dependencies (for any build system), it means the build system doesn't have all the appropriate dependencies and so isn't building everything needed for the failing build step.  I run into this often (though mainly with non-scons based build systems where it's harder to get this right)..

**Comment by dnwillia-work** [comment]
Yeah, fair enough.  Even for the parallel builds I'll add that it would sometimes work, sometimes not.

I wonder since it seems to be a windows specific thing then is something else going on.  eg: the .mod file is still locked by the OS so the dependent module cannot link against it.  We have run into this kind of issue with .lib (import library) files on windows parallel builds before as well if an `Install()` was used to move the .lib file to some other place before it was linked into dependent module despite the fact that we would add `Depends()` on the install task to the dependent module.  We removed the use of `Install()` targets for this as a result and it got better.

**Comment by bdbaddog** [comment]
Ugh.. the joys of NTFS..

**Comment by bdbaddog** [comment]
How are you creating your Environment()? Specifically are you setting `FORTRANSUFFIXES` (or any of the `F*SUFFIXES` env vars)?

**Comment by bdbaddog** [comment]
Try adding this to your Environment()

`                   FORTRANFILESUFFIXES=['.f90'], `

**Comment by james-thunes** [comment]
this particular library has all *.f files, but I can try setting `FORTRANFILESUFFIXES=['.f'],`

**Comment by dnwillia-work** [comment]
> How are you creating your Environment()? Specifically are you setting `FORTRANSUFFIXES` (or any of the `F*SUFFIXES` env vars)?

We definitely do *not* configure that particular variable currently.  We do configure these variables related to this issue
```
    env["FORTRANMODDIR"] = "${TARGET.dir}"
    env["F90PATH"] = "${TARGET.dir}"
```
It's the same for `ifort` and `gfortran`.

**Comment by mwichmann** [comment]
Okay, I've located a simple example and don't see the problem with it. Here's the tree dump:
```console
+-.
  +-SConstruct
  +-list-module.f90
  +-list-module.o
  | +-list-module.f90
  | +-/bin/gfortran
  +-listmodule.mod
  | +-list-module.f90
  | +-/bin/gfortran
  +-main-list
  | +-main-list.o
  | | +-main-list.f90
  | | +-[listmodule.mod]
  | | +-/bin/gfortran
  | +-[list-module.o]
  | +-/bin/gfortran
  +-main-list.f90
  +-[main-list.o]
```
`main-list.o` is shown as depending on its source file, on `listmodule.mod`, and on the compiler.  That looks right to me? What am I missing?

**Comment by bdbaddog** [comment]
@mwichmann - I think this is only an intel fortran issue. And I think it's due to loading default tools, then ifort and having files named .f when ifort looks like it only wants files named .i or .i90?

**Comment by bdbaddog** [comment]
> this particular library has all *.f files, but I can try setting `FORTRANFILESUFFIXES=['.f'],`

Did that work?

**Comment by mwichmann** [comment]
I still have to look at the .i/.i90 thing, I think it should work just find on ordinary suffixes and I don't know what those actually mean.

**Comment by james-thunes** [comment]
> > this particular library has all *.f files, but I can try setting `FORTRANFILESUFFIXES=['.f'],`
> 
> Did that work?

Sorry, been busy with other tasks. Will try to get to this as soon as I can.

**Comment by james-thunes** [comment]
> > this particular library has all *.f files, but I can try setting `FORTRANFILESUFFIXES=['.f'],`
> 
> Did that work?

Ok, I was able to spend a bit of time with this today. I'm a bit confused to be honest. Setting `FORTRANFILESUFFIXES=['.f'],` did indeed appear to resolve the issue. I was able to see the module dependencies when looking at the output with `--tree=prune`. However, removing the flag and doing a clean rebuild of the code I see that the module dependencies is still seen. It definitely wasn't last week...

**Comment by mwichmann** [comment]
Most of this setup bemuses me, so you have company :)

Out of curiosity, are all the files in your project of the same suffix?

**Comment by dnwillia-work** [comment]
The library in question has a combination of both `.f90` and `.f` suffixes since it includes F90 free format and F77 fixed format code.

**Comment by mwichmann** [comment]
Okay, then the muck with "dialects" isn't completely useless to you (or if it is - please feel free to let us know!).

**Comment by dnwillia-work** [comment]
Right, the library in question uses both dialects because it links in an older code base shared with other applications.  So, definitely useful to us.  One issue though is that it's still valid to have F90/95 etc... code in fixed form files with a `.f` extension, so it muddies the water some more.

I tried the `--tree=prune` thing with the latest SCons.
`
SCons by Steven Knight et al.:
        SCons: v4.3.0.559790274f66fa55251f5754de34820a29c7327a, Tue, 16 Nov 2021 19:09:21 +0000, by bdeegan on octodog
`

on the problematic code and I get the following
```
  | +-<path-to-lib>\importLibrary.lib
  |   +-<path-obj>\module1.obj
  |   | +-<path-src>\module1.f
  |   | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  |   +-<path-obj>\module2.obj
  |   | +-<path-src>\module2.f
  |   | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  |   +-<path-obj>\fortranFile1.obj
  |   | +-<path-src>\fortranFile1.f
  |   | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
```
In this instance, `fortranFile1.f` has both `use module1` and `use module2` statements within the contained functions/subroutines, yet the dependency is not showing up in the tree.  However, if I look further down in the tree I see this:
```
  |   +-<path-obj>\fortranFile2.obj
  |   | +-<path-src>\fortranFile2.f90
  |   | +-<path-obj>\moduleA.mod
  |   | | +-<path-src>\moduleA.f90
  |   | | +-<path-obj>\moduleB.mod
  |   | | | +-<path-src>\moduleB.f90
  |   | | | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  |   | | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  |   | +-[<path-obj>\moduleB.mod]
  |   | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
```
So, it seems to be getting the dependency correct for `.f90` extensions but not `.f` extensions. 

Could it be that the `.mod` files are not added here properly when we are passing `.f` files that use modules to the library target?

**Comment by mwichmann** [comment]
> So, it seems to be getting the dependency correct for .f90 extensions but not .f extensions.

That's not impossible, since they use separate settings groups, but this stuff is convoluted enough it's hard to see. There are also slightly differing settings for lib and non-lib.

Did you say you were using a modified copy of ifort.py?

Is it possible to attach a dump of the environment variables (really only the set that start with *F* are interesting, I believe) - it *might* show the anomaly that's hitting your usage.

**Comment by dnwillia-work** [comment]
Yes, we have a modified version of ifort.py.  Staring at the differences with your current code the diffs are:

- It configures msvc tool if needed. This could maybe be removed.
- We have a few functions that hunt for the installed location of the compiler, gets all the versions etc... and then picks the latest.  I think I got this code from your project at some point, it seems to have been disappeared.
- Sets the variables FORTRANMODDIR and F90PATH as I mentioned earlier

Here's the variables:
```py
FORTRANFLAGS = ['/nologo', '/MD', '/Qvc14', '/warn:unused', '/warn:uncalled', '/O3'] # We set this one
F90FLAGS = ['/nologo', '/MD', '/Qvc14', '/warn:unused', '/warn:uncalled', '/O3'] # We set this one
F95FLAGS = ['/nologo', '/MD', '/Qvc14', '/warn:unused', '/warn:uncalled', '/O3'] # We set this one
FORTRANPATH = ['#.']
FORTRANSUFFIXES = ['.f', '.for', '.ftn', '.F', '.FOR', '.FTN', '.fpp', '.FPP', '.f77', '.F77', '.f90', '.F90', '.f95', '.F95', '.f03', '.F03', '.f08', '.F08']
FORTRANCOM = $FORTRAN -object:$TARGET -c $FORTRANFLAGS $_FORTRANINCFLAGS $_FORTRANMODFLAG $SOURCES
FORTRANPPCOM = $FORTRAN -object:$TARGET -c $FORTRANFLAGS $CPPFLAGS $_CPPDEFFLAGS $_FORTRANINCFLAGS $_FORTRANMODFLAG $SOURCES
FORTRANMODPREFIX =
FORTRANMODSUFFIX = .mod
FORTRANMODDIR = ${TARGET.dir} # We set this one
FORTRANMODDIRPREFIX = /module:
FORTRANMODDIRSUFFIX =
F77FLAGS =
F77COM = $F77 -object:$TARGET -c $F77FLAGS $_F77INCFLAGS $SOURCES
F77PPCOM = $F77 -object:$TARGET -c $F77FLAGS $CPPFLAGS $_CPPDEFFLAGS $_F77INCFLAGS $SOURCES
F90COM = $F90 -object:$TARGET -c $F90FLAGS $_F90INCFLAGS $_FORTRANMODFLAG $SOURCES
F90PPCOM = $F90 -object:$TARGET -c $F90FLAGS $CPPFLAGS $_CPPDEFFLAGS $_F90INCFLAGS $_FORTRANMODFLAG $SOURCES
F95COM = $F95 -object:$TARGET -c $F95FLAGS $_F95INCFLAGS $_FORTRANMODFLAG $SOURCES
F95PPCOM = $F95 -object:$TARGET -c $F95FLAGS $CPPFLAGS $_CPPDEFFLAGS $_F95INCFLAGS $_FORTRANMODFLAG $SOURCES
F03FLAGS =
F03COM = $F03 -o $TARGET -c $F03FLAGS $_F03INCFLAGS $_FORTRANMODFLAG $SOURCES
F03PPCOM = $F03 -o $TARGET -c $F03FLAGS $CPPFLAGS $_CPPDEFFLAGS $_F03INCFLAGS $_FORTRANMODFLAG $SOURCES
F08FLAGS =
F08COM = $F08 -o $TARGET -c $F08FLAGS $_F08INCFLAGS $_FORTRANMODFLAG $SOURCES
F08PPCOM = $F08 -o $TARGET -c $F08FLAGS $CPPFLAGS $_CPPDEFFLAGS $_F08INCFLAGS $_FORTRANMODFLAG $SOURCES
F77 = ifort
F90 = ifort
FORTRAN = ifort
F95 = ifort
F90PATH = ${TARGET.dir}  # We set this one
```
Noted which ones we explicitly set.

**Comment by dnwillia-work** [comment]
Based on my hypothesis I tried to reproduce this with a simple example, could not do it yet....

**Comment by bdbaddog** [comment]
@dnwillia-work - could you provide your modified version of ifort.py ?

**Comment by dnwillia-work** [comment]
Yeah sure. I’ll push it up later today and share a PR.

**Comment by mwichmann** [comment]
At least at my end, the effort of the `ifort.py` tool to add those extra recognized suffixes (`.i` and `.i90`) messes things up - you get into a place where apparently the `.f` files aren't scanned (the `.f90` ones aren't either, so this doesn't explain your issue).  Are those actually used for anything?  As I said... somewhere... I can't find any hint of those in current Intel docs, which aren't particularly searchable.

**Comment by dnwillia-work** [comment]
Regarding the file extensions there is this: 

https://www.intel.com/content/www/us/en/develop/documentation/fortran-compiler-oneapi-dev-guide-and-reference/top/compiler-setup/use-the-command-line/file-extensions.html

So, Intel does seem to accept `.i` and `.i90` as acceptable file extensions where those files are passed to the compiler.

**Comment by mwichmann** [comment]
Sigh... I'd swear I had been looking at that very page, must be blind.

**Comment by dnwillia-work** [comment]
lol.  Yeah, it's not the easiest to navigate or digest it.

Here's the ifort tool module we are using currently:

https://github.com/dnwillia-work/scons/pull/1

**Comment by mwichmann** [comment]
There are these comments [in the wiki](/scons/scons/wiki/FortranDiscussion)

**Comment by dnwillia-work** [comment]
Thanks for sharing that.  In general, I agree with what is proposed.  The stuff about the dialects is probably overly complex.

I tried again to reproduce this issue outside our regular build system using the ifort.py tool I shared and cannot.  My SConstruct looks like this: 
```py
import os
localTools = os.path.join('.', 'site_scons', 'site_tools')
env = Environment(toolPath = localTools, tools = ['default', 'ifort'])
env['LINKFLAGS'] += ['/subsystem:console'] # Needed to link a Program()
env.Program('hello77', ['hello77.f', 'fixedModule.f', 'freeModule.f90'])
env.Program('hello90', ['hello90.f90', 'fixedModule.f', 'freeModule.f90'])
```
Source files are:

hello77.f
```fortran
      program hello
      use fixedModule, only : sayHello
      use freeModule, only : sayHello90
      call sayHello
      call sayHello90
      end program hello
```
hello90.f90
```fortran
program hello90
  use fixedModule, only : sayHello
  use freeModule, only : sayHello90
  call sayHello
  call sayHello90
end program hello90
```
freeModule.f90
```fortran
module freeModule
  public :: sayHello90
contains
  subroutine sayHello90()
    print *,"Hello from free form module!"
  end subroutine sayHello90
end module freeModule
```
fixedModule.f
```fortran
      module fixedModule
      public :: sayHello
      contains
      subroutine sayHello()
        print *,"Hello from fixed form module!"
      end subroutine sayHello
      end module fixedModule
```
I get this tree:
```
+-.
  +-fixedModule.f
  +-fixedmodule.mod
  | +-fixedModule.f
  | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  +-fixedModule.obj
  | +-fixedModule.f
  | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  +-freeModule.f90
  +-freemodule.mod
  | +-freeModule.f90
  | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  +-freeModule.obj
  | +-freeModule.f90
  | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  +-hello77.exe
  | +-hello77.obj
  | | +-hello77.f
  | | +-[fixedmodule.mod]
  | | +-[freemodule.mod]
  | | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  | +-[fixedModule.obj]
  | +-[freeModule.obj]
  | +-C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.20.27508\bin\HostX64\x64\link.EXE
  +-hello77.f
  +-[hello77.obj]
  +-hello90.exe
  | +-hello90.obj
  | | +-hello90.f90
  | | +-[fixedmodule.mod]
  | | +-[freemodule.mod]
  | | +-C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\bin\intel64\ifort.EXE
  | +-[fixedModule.obj]
  | +-[freeModule.obj]
  | +-C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.20.27508\bin\HostX64\x64\link.EXE
  +-hello90.f90
  +-[hello90.obj]
  +-SConstruct
```
So, it all looks ok.

Interesting side point: I have to add `/subsystem:console` there for `env.Program()` targets to link otherwise the linker does not find the entry point.

**Comment by dnwillia-work** [comment]
@mwichmann I ran into this again today on a fairly simple project I'm messing with.

I did some tracing/debugging and what I found is that this warning gets generated if I pass `-warn=all` on the command line:

https://github.com/SCons/scons/blob/810ca6c8895b01cbd636d83079f6a848dc36adf6/SCons/Scanner/Fortran.py#L114

when a module uses another module.  So, if you do this in my prior example:
```
      module fixedModule
      use freeModule
      public :: sayHello
      contains
      subroutine sayHello()
        print *,"Hello from fixed form module!"
      end subroutine sayHello
      end module fixedModule
```
then that fails because `fixedModule` is compiled without compiling `freeModule` first.

I checked that the scanner perfectly figures out that `freeModule.mod` is a dependency of `fixedModule.f` but `n` here is returned as none:
```
n, i = self.find_include(dep, source_dir, path)
```
and you get the warning.  I find it hard to parse everything that goes on if I follow the rabbit hole down the function call to `find_include` but if the general assumption is that the module files will be present in the source tree like .h, .c, .f, .f90 etc... then that assumption is incorrect.  The module files are a compiler output file, i.e. generated by the compiler.

**Comment by mwichmann** [comment]
Interesting. Is this only with ifort?

**Comment by bdbaddog** [comment]
> @mwichmann I ran into this again today on a fairly simple project I'm messing with.
> 
> I did some tracing/debugging and what I found is that this warning gets generated if I pass `-warn=all` on the command line:
> 
> https://github.com/SCons/scons/blob/810ca6c8895b01cbd636d83079f6a848dc36adf6/SCons/Scanner/Fortran.py#L114
> 
> when a module uses another module. So, if you do this in my prior example:
> 
> ```
>       module fixedModule
>       use freeModule
>       public :: sayHello
>       contains
>       subroutine sayHello()
>         print *,"Hello from fixed form module!"
>       end subroutine sayHello
>       end module fixedModule
> ```
> 
> then that fails because `fixedModule` is compiled without compiling `freeModule` first.
> 
> I checked that the scanner perfectly figures out that `freeModule.mod` is a dependency of `fixedModule.f` but `n` here is returned as none:
> 
> ```
> n, i = self.find_include(dep, source_dir, path)
> ```
> 
> and you get the warning. I find it hard to parse everything that goes on if I follow the rabbit hole down the function call to `find_include` but if the general assumption is that the module files will be present in the source tree like .h, .c, .f, .f90 etc... then that assumption is incorrect. The module files are a compiler output file, i.e. generated by the compiler

Can you paste output when run with `scons --tree=all` ?

**Comment by mwichmann** [comment]
> The module files are a compiler output file, i.e. generated by the compiler.

As to the last sentence - at least on some level, the code knows that, that's what emitters do, add in files that are generated by tools.

**Comment by mwichmann** [comment]
Just as a note (unrelated to the issue), this stanza:
```py
localTools = os.path.join('.', 'site_scons', 'site_tools')
env = Environment(toolPath = localTools, tools = ['default', 'ifort'])
```
should be unneeded - SCons already looks in `site_scons/site_tools` without you having to add that to the `toolpath`.

**Comment by mwichmann** [comment]
and I can run this example, assuming I got the whole thing including problems, with gfortran (and taking out the Windows-specific flag), so I'm assuming we'll be able to spot what's different in the setup evenutally.

**Comment by dnwillia** [comment]
@mwichmann Sorry for cluttering up this thread.  I did a bunch more investigation and have reproduced a few issues, including the one in this thread which looks like it is due to Fortran code with modules not working right with `variant_dir`  See this repo:

https://github.com/dnwillia/SConsTests

I think everything for this issue is reproduced there on a really basic example.

**Comment by mwichmann** [comment]
Okay.  We admittedly have a lot of Fortran issues already, and not enough people using it to drive getting things fixed.  Maybe this is a moment.

https://github.com/SCons/scons/labels/Fortran

#2635 (and maybe a few others) look like they may be related - or not; that was just a really quick glance.

**Comment by dnwillia** [comment]
lol, yes.  James and I were speculating last night we are the only ones.

**Comment by bdbaddog** [comment]
Are you running of the master development branch? Or the latest release (4.5.2)?

**Comment by mwichmann** [comment]
From the linked gh:

> None of this code builds out of the box at all on Windows. The ifort tool does not configure properly even if you have it pre-configured in the environment.

This is no surprise. The current Intel opeAPI toolset depends heavily on a setup that fills in environment variables, which SCons proceeds to ignore in its search for purity and reproducibility. Things which are vital to actually running ifort, icc, and their next-gen equivalents need to bet set up in the tool.  The msvc setup tool has the same problem, and it's grown substantially over the years are more and more stuff becomes needed.  I know I've said this before, sorry if it's repeating: this kind of heavyweight (you count on a few thousand lines of shell script or batch file to have run, leaving your environment ready-to-go) isn't a great fit for the current SCons model of simple tool modules and isolation of the environment in which the builds actually run - it makes for a lot of work, and the ground keeps changing under you.

**Comment by dnwillia** [comment]
@bdbaddog The tests are all with 4.5.2... I've not quite figured out how to create a wheel yet from your repo.  Happy to try that.

**Comment by dnwillia** [comment]
> 

Matt, regarding Intel compilers I think the compromise would be to have the tool call the necessary configuration script for the Intel compiler to setup the Environment object so that code will build.  I have a custom config tool that works, which I derived from a SCons provided tool years ago,  but it's kind of a pain to maintain it.

**Comment by bdbaddog** [comment]
> @bdbaddog The tests are all with 4.5.2... I've not quite figured out how to create a wheel yet from your repo. Happy to try that.

If you have a checked out git repo, you can just run via `<path to your checked out scons>/scripts/scons.py <your normal args here>`

**Comment by mwichmann** [comment]
Okay, and this line:

> Building code with fortran modules does not work right with variant_dir. You need to set the location where fortran modules (.mod) get generated otherwise they get stored into the same directory as the SConstruct.

Yes, this comes up repeatedly.  SConscripts are evaluated in the context of the directory they're in, but then the builds happen in the context of the SConstruct dir (simplification, but will do for now).   Just had a wrestling match with flex/bison "extra files" over this.  I know the past Fortran tool maintainers knew this, but that doesn't mean every combination is right. Certainly when glancing back over some of the past bugs this theme comes up several times - if there's a subdirectory involved (which there *always* is if there's a varintdir in use), then things can go wrong.

**Comment by mwichmann** [comment]
> > @bdbaddog The tests are all with 4.5.2... I've not quite figured out how to create a wheel yet from your repo. Happy to try that.
> 
> If you have a checked out git repo, you can just run via `<path to your checked out scons>/scripts/scons.py <your normal args here>`

Or, in the top directory of your git checkout of scons, and with your virtualenv activated: 
```py
pip install -e .
```

**Comment by dnwillia** [comment]
OK let me give it a shot.  I did not think of that.

**Comment by dnwillia** [comment]
Alright, done, updated the README with some notes about the install into my environment.

All the issues persist, including this one.

Matt I think the behaviour you describe here:

> Yes, this comes up repeatedly. SConscripts are evaluated in the context of the directory they're in, but then the builds happen in the context of the SConstruct dir (simplification, but will do for now). Just had a wrestling match with flex/bison "extra files" over this. 

is fine.  It's just that the object files produced by the compiler do not end up in that directory.  The module files should really be going to the same place as the object files.  There is a link in my notes to the change I made that gets them into the right spot.

**Comment by mwichmann** [comment]
Which, at first glance, sounds sane (if you mean `env["FORTRANMODDIR"] = "${TARGET.dir}"`). Waiting on @bdbaddog though!

**Comment by dnwillia** [comment]
> Which, at first glance, sounds sane (if you mean `env["FORTRANMODDIR"] = "${TARGET.dir}"`). Waiting on @bdbaddog though!

Yes, that's the line I mean.

**Comment by bdbaddog** [comment]
Do you always want the .mod files in the same directory as the .o's ?

**Comment by dnwillia** [comment]
Good question.  I would say possibly not, but it's certainly a much better default than what happens at the moment (i.e. having them dumped in the cwd with the SConstruct) and will get the `variant_dir` builds working properly.

**Comment by mwichmann** [comment]
@dnwillia it seems like your repository of tests has differing behavior between runs. I tried `SConstructVariant` once and saw the error you saw. Edited - I misread, this looks like a different error...
```console
gfortran -o TestSConscript/builds1/linux/main_prog.exe TestSConscript/builds1/linux/module_one.o TestSConscript/builds1/linux/module_two.o TestSConscript/builds1/linux/main_prog.o
scons: *** [TestSConscript/builds2/linux/builds2] TypeError : Tried to lookup Dir 'TestSConscript/builds2' as a File.
```
then tried it again and the build worked.  (!?!? I hate those). Is this maybe some case of the sconsign file getting populated?

@bdbaddog
> Do you always want the .mod files in the same directory as the .o's ?

Thinking about it, I believe not, unless it's a fairly flat tree.  If you `USE` a module, then its `.mod` file has to be available at compilation time. If the source file being compiled is in a different directory than the module's source, then having the module file over in that other directory isn't any good either unless `FORTRANMODDIR` points over there. 
```
       -Jdir
           This option specifies where to put .mod files for compiled modules. 
           It is also added to the list of directories to searched by an "USE" statement.

           The default is the current directory.
```
So if we do put them together with the module's objects, then that directory should somehow get added to `FORTRANPATH` so it gets searched later:
```
       -Idir
           ...
           This path is also used to search for .mod files when previously compiled modules are required
           by a "USE" statement.
```
Can we leave that step up to the SConscript file author?

**Comment by dnwillia** [comment]
Regarding the first issue, yes I think I ran into that as well.  Not sure why.  The example SCons input files are not particularly complicated and are based on what is written into the user documentation for SCons so I'm just following the pattern.  :-)

Regarding the .mod location I think it just comes down to what you want the _default_ behavior to be.  The default right now is to leave them in the SConstruct directory when using the `variant_dir` feature.  Without the change to `FORTRANMODDIR` the `--warn=all` option reports warnings about the .mod files not being found (as I noted already). I think that the .mod files are an _output_ of the build process and so should be in the `variant_dir` if you use that as in my repo.  Maybe my solution is not the right way to make this happen, but this would be my expected UX.

I agree that if you want to modify this behavior then of course you need to resort to using `-J` and `-I` and/or modifying the `FORTRANMODDIR` variable.

**Comment by bdbaddog** [comment]
> @dnwillia it seems like your repository of tests has differing behavior between runs. I tried `SConstructVariant` once and saw the error you saw. Edited - I misread, this looks like a different error...
> 
> ```
> gfortran -o TestSConscript/builds1/linux/main_prog.exe TestSConscript/builds1/linux/module_one.o TestSConscript/builds1/linux/module_two.o TestSConscript/builds1/linux/main_prog.o
> scons: *** [TestSConscript/builds2/linux/builds2] TypeError : Tried to lookup Dir 'TestSConscript/builds2' as a File.
> ```
> 
> then tried it again and the build worked. (!?!? I hate those). Is this maybe some case of the sconsign file getting populated?
> 
> @bdbaddog
> 
> > Do you always want the .mod files in the same directory as the .o's ?
> 
> Thinking about it, I believe not, unless it's a fairly flat tree. If you `USE` a module, then its `.mod` file has to be available at compilation time. If the source file being compiled is in a different directory than the module's source, then having the module file over in that other directory isn't any good either unless `FORTRANMODDIR` points over there.
> 
> ```
>        -Jdir
>            This option specifies where to put .mod files for compiled modules. 
>            It is also added to the list of directories to searched by an "USE" statement.
> 
>            The default is the current directory.
> ```
> 
> So if we do put them together with the module's objects, then that directory should somehow get added to `FORTRANPATH` so it gets searched later:
> 
> ```
>        -Idir
>            ...
>            This path is also used to search for .mod files when previously compiled modules are required
>            by a "USE" statement.
> ```
> 
> Can we leave that step up to the SConscript file author?

Nope. Fairly certain that means the DAG isn't right, and once the build fails, the needed files (which Scons isn't aware of) are in the right place to succceed..
Check scons --tree=all ?

**Comment by mwichmann** [comment]
Here's a dump, I have to run off and do other stuff....

[Variant-output.txt](https://github.com/SCons/scons/files/11908256/Variant-output.txt)

**Comment by bdbaddog** [comment]
> Variant-output.txt

Ugh. that's odd.

**Comment by mwichmann** [comment]
Had a bit of a dig, and it's just too convoluted and too many layers to tease out much info.  For the case of non-duplicating variantdir, we die in the Fortran scanner, in `scan`, where it:
```py
        for dep in mods_and_includes:                                          
            n, i = self.find_include(dep, source_dir, path)  
```
and doesn't find the module it's looking for... it looks like it's looking for them as if they were static files that will be present in the source tree?  but there must be more going on here since the mod files are derived, not static sources, and this comes out correct in other usages.   Anyway, with verbose enabled in the file finder, first the copying-variant form:
```
scons: Building targets ...
  find_file: looking for 'module_two.mod' in 'TestSConscript/builds1/linux' ...
  find_file: ... FOUND 'module_two.mod' in 'TestSConscript/builds1/linux'
gfortran -o TestSConscript/builds1/linux/module_two.o -c -JTestSConscript/builds1/linux TestSConscript/builds1/linux/module_two.f90
gfortran -o TestSConscript/builds1/linux/module_one.o -c -JTestSConscript/builds1/linux TestSConscript/builds1/linux/module_one.f90
  find_file: looking for 'module_one.mod' in 'TestSConscript/builds1/linux' ...
  find_file: ... FOUND 'module_one.mod' in 'TestSConscript/builds1/linux
```
vs the non-copying variant:
```
scons: Building targets ...
  find_file: looking for 'module_two.mod' in 'TestSConscript' ...

scons: warning: No dependency generated for file: module_two.mod (referenced by: TestSConscript/module_one.f90) -- file not found
File "/home/mats/.pyenv/versions/venv-system/bin/scons", line 8, in <module>
gfortran -o TestSConscript/builds12/linux/module_one.o -c -JTestSConscript/builds12/linux TestSConscript/module_one.f90
TestSConscript/module_one.f90:2:7:

    2 |   use module_two
      |       1
Fatal Error: Cannot open module file 'module_two.mod' for reading at (1): No such file or directory
```
Notice the different path the file finder is passed.

On a separate note, when the revised `gfortran.py` from the test branch sets up `env["FORTRANMODDIR"] = "${TARGET.dir}"`, that doesn't get expanded as expected *in the emitter* when it's called, the emitter does this:
```py
moddir = env.subst('$FORTRANMODDIR', target=target, source=source)
```
and always comes back here with `moddir` set to `.`; the expansion works right when the command line is actually generated as can be seen above.

**Comment by dnwillia-work** [comment]
> and doesn't find the module it's looking for... it looks like it's looking for them as if they were static files that will be present in the source tree? 

That was what I managed to work out too but from there it's not clear to me what to do about it so I stopped digging.

**Comment by mwichmann** [comment]
Okay, a little more light on the non-copying variantdir: sources are absolute paths in this case, but targets shouldn't be so they can be handled by the variant code. But the module that will be generated, as detected by the emitter (not via the scanner, btw, but by using a regex that's supposed to stay in sync with the one in the scanner), has the absolute-path form.

```
Fortran emitter: module_one.f90 -> module_one.o module_one.mod
Fortran emitter: module_two.f90 -> module_two.o module_two.mod
Fortran emitter: main_prog.f -> main_prog.o
...
Fortran emitter: /home/mats/github/scons/exp/fort-subdir/SConsTests/TestSConscript/module_one.f90 -> module_one.o /home/mats/github/scons/exp/fort-subdir/SConsTests/TestSConscript/module_one.mod
Fortran emitter: /home/mats/github/scons/exp/fort-subdir/SConsTests/TestSConscript/module_two.f90 -> module_two.o /home/mats/github/scons/exp/fort-subdir/SConsTests/TestSConscript/module_two.mod
Fortran emitter: /home/mats/github/scons/exp/fort-subdir/SConsTests/TestSConscript/main_prog.f -> main_prog.o
```

So I presume that this snip from FortranCommon is the wrong way to generate the target node(s) for modules:
```py
    for module in modules:                                                     
        target.append(env.fs.File(module, moddir))
```

**Comment by mwichmann** [comment]
If it helps to be explicit - it's not just that the module paths here are absolute, it's that they're to what would be considered the "source directory", as in:

*absolute-path-to-topdir*`/TestSConscript/module_one.mod`

the module file will never appear there... it will either be in the topdir if there's no `FORTRANMODDIR`, or in the build directory specified by the variantdir if it is set as in the sample.

**Comment by mwichmann** [comment]
Okay, I'm just repeating stuff that's already been said, but module handling seems pretty broken. I set up another small test with a trivial module source file, repeated four times (uniqued for each version):

1. source in top-level directory
2. source in `src` subdirectory with own SConscript
3. source in `vsrc` subdirectory, SConscript called as duplicating variantdir `build`
4. source in `vsrc-no` subdirectory, SConscript called as non-duplicating variantdir `build-no`

Calling `Object()` on each, with no setting of `FORTRANMODDIR`, so the default empty string, which won't expand to any `-J` option to `gfortran`, you expect to get four object files and four `.mod` files, and so we do:
```
build-no/test4.o
build/test3.o
mod_foo1.mod
mod_foo2.mod
mod_foo3.mod
mod_foo4.mod
src/test2.o
test1.o
```
But that doesn't match what SCons thinks it did, which means we have a problem (the tree is trimmed to show only these eight):
```
+-.
  +-build
  | +-build/mod_foo3.mod
  | +-build/test3.o
  +-build-no
  | +-build-no/mod_foo4.mod
  | +-build-no/test4.o
  +-mod_foo1.mod
  +-src
  | +-src/mod_foo2.mod
  | +-src/test2.o
  +-test1.o
```
It's never good when SCons builds up a picture of what exists, and then that's not what the compiler did.  The *simplest* problem from that is you can't clean, but there's plenty more issues, like rebuilds being horked, etc.

**Comment by mwichmann** [comment]
Now, if we redo the above with the experimental setting of `FORTRANMODDIR`, so it's force-passed on every build, we at least get consistency:
```
f95 -o build/test3.o -c -Jbuild build/test3.f
f95 -o build-no/test4.o -c -Jbuild-no vsrc-no/test4.f
f95 -o test1.o -c -J. test1.f
f95 -o src/test2.o -c -Jsrc src/test2.f
```
Giving us:
```
build/mod_foo3.mod
build/test3.o
build-no/mod_foo4.mod
build-no/test4.o
mod_foo1.mod
src/mod_foo2.mod
src/test2.o
test1.o
```
And the tree at least now matches what was built:
```
+-.
  +-build
  | +-build/mod_foo3.mod
  | +-build/test3.o
  +-build-no
  | +-build-no/mod_foo4.mod
  | +-build-no/test4.o
  +-mod_foo1.mod
  +-src
  | +-src/mod_foo2.mod
  | +-src/test2.o
  +-test1.o
```
But that hasn't got into the other case of trying to `USE` these modules, and whether they're found where they got put so other sources will build.

**Comment by dnwillia** [comment]
Yup. At least you can reproduce what I’m also seeing.

**Comment by mwichmann** [comment]
I thought I had a cute fix for the "if there's no `FORTRANMODDIR`, SCons is out of sync with what the compiler does" problem, but it turns out an extra bit of variantdir cleverness defeats that in one case. Still thinking.  This was the idea:
```py
    for module in modules:
        if not moddir:
           target.append(env.fs.File(module, directory='#'))
        else:
           target.append(env.fs.File(module, directory=moddir))
```

**Comment by mwichmann** [comment]
So the odd bit is recorded, it seems, if you set a relative path for `FORTRANMODDIR`, and use a non-duplicating variantdir, scons will double the flag usage, emitting one module dir for the target (variant) directory and one for the original source directory, and that's illegal.  Like

```py
env = Environment(FORTRANMODDIR="modules")
```

which leads to:
```
gfortran -o build-no/test4.o -c -Jbuild-no/modules -Jvsrc-no/modules vsrc-no/test4.f
Fatal Error: gfortran: Only one '-J' option allowed
compilation terminated.
```

I don't think I've caused this....

Leaving it out (the default is an empty string), or using a top-relative path (`"#/modules"`) does not cause this effect.

**Comment by mwichmann** [comment]
If a relative path is used, then `RDirs`, the function passed in the `_concat` call, is the one that makes this mess.

So this:
```py
env['_FORTRANMODFLAG'] = '$( ${_concat(FORTRANMODDIRPREFIX, FORTRANMODDIR, FORTRANMODDIRSUFFIX, __env__, RDirs, TARGET, SOURCE)} $)'
```
has the prospect of producing something illegal, with Fortran compilers not accepting multiple module-directory options. I think it's better not to apply this particular transformation (i.e., leave out `RDirs` from the generation of this flag.)

**Comment by mwichmann** [comment]
Hoo boy, this simple little matter might end up being a hair-loss event...

**Comment by mwichmann** [comment]
> On a separate note, when the revised `gfortran.py` from the test branch sets up `env["FORTRANMODDIR"] = "${TARGET.dir}"`, that doesn't get expanded as expected _in the emitter_ when it's called, the emitter does this:
> 
> ```python
> moddir = env.subst('$FORTRANMODDIR', target=target, source=source)
> ```
> 
> and always comes back here with `moddir` set to `.`; the expansion works right when the command line is actually generated as can be seen above.

Okay, this conclusion was just bogus.  At SConscript evaluation time, the context is the directory of the SConscript, so it makes sense for the `subst` here to come back with `.`. Ignore me...

**Comment by dnwillia** [comment]
OK so an update to this finally... Today I tried out all my tests posted at 

https://github.com/dnwillia/SConsTests

and everything is building properly at release `v4.8.1` on MacOS using gfortran.  Great progress.

I'll give it a shot with Linux/windows with both Intel (Windows/Linux) and GNU fortran (Linux) next week.  @james-thunes

**Comment by james-thunes** [comment]
great to hear. Thanks for giving it another try. And thanks to the developers for addressing :)

**Comment by dnwillia** [comment]
OK so I would say I can confirm that on Linux and MacOS that the code in my test repo compiles with v4.8.1 out of the box with gfortran on Linux and MacOS.  It's also working on windows up to the linker step, where linking fails, but that's a separate issue from this one.

@james-thunes Maybe you can close this issue unless you still see problems.
